### DISCUSSION TITLE
How do I use slf4j 2 and slf4j 1.7 at the same time?

### DISCUSSION THREAD

**Discussion by rupert-madden-abbott** [qos-ch/slf4j#319] [Q&A]
I've just updated a library which has in turn updated to slf4j 2. Other libraries I am using are still depending on slf4j 1.7.

The result of this is my logging has broken and the following message is output at startup:

```
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions 1.7.x or earlier.
SLF4J: Ignoring binding found at [jar:file:/.m2/repository/ch/qos/logback/logback-classic/1.2.11/logback-classic-1.2.11.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.
```

An added complication for me is that the library I am depending on, `wiremock-jr8-standalone` has shaded the slf4j 2 API so there was no hint in my dependency tree that slf4j 2 was getting pulled in.

Anyway, is the above behaviour anticipated and, if so, what is the intended upgrade path? I am supposed to keep everything on slf4j 1.7 until everything can move to 2.0? Or am I supposed to remove all 1.7 bindings and replace them with 2.0 providers if at least one dependency is on 2.0? Is it the shading that has caused the problem specifically somehow?

**Comment by ceki** [comment]
Hi @rupert-madden-abbott,

Have you read https://www.slf4j.org/codes.html#ignoredBindings mentioned in the warning message?  Does it shed any light onto the matter?

**Reply by rupert-madden-abbott** [reply]
Hi @ceki,

Thanks for responding!

Yes I did find that as it is very handily linked to from the error logs (thanks for that!).

I think the problem is due to the shading. I have both slf4j 1.7.x and slf4j 2.x on my classpath simultaneously which would not normally be possible right? The group ID, artifact ID and packages have all not changed from 1.7.x to 2.x.

So I think the surprise for me is in introducing a shaded slf4j 2.x, I broke logging that appears to be going via the slf4j-api 1.7.x.

**Comment by ceki** [comment]
The warning message mentioning "provider" means that SLF4J 2.0.x is being loaded, and not SLF4J 1.7.x. I don't know your set up but in principle placing a compatible logging implementation such as logback 1.3 or later will solve the problem. What is the logging implementation you are using currently?
